### PR TITLE
Make codecharta reload save #2056

### PR DESCRIPTION
**Pull request by fritschldwg**
# Make codecharta reload save #2056

Closes: #2056

## Description
  - Previous state (uploaded files and current configuration) is restored upon reload

## Definition of Done

A PR is only ready for merge once all the following acceptance criteria are fulfilled:
- [x] Changes have been manually tested
- [x] All TODOs related to this PR have been closed
- [x] There are automated tests for newly written code and bug fixes
- [x] All bugs discovered while working on this PR have been submitted as issues (if not already an open issue)
- [x] Documentation (GH-pages, analysis/visualization READMEs, parser READMEs, --help, etc.) has been updated (almost always necessary except for bug fixes)
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: visualization/package.json [hunk 1]
```diff
@@ -115,11 +116,13 @@
 		"@typescript-eslint/eslint-plugin": "^5.44.0",
 		"@typescript-eslint/parser": "^5.44.0",
 		"bestzip": "^2.2.0",
+		"core-js": "^3.35.1",
 		"electron-packager": "^17.1.2",
 		"eslint": "^8.28.0",
 		"eslint-config-prettier": "^8.5.0",
 		"eslint-plugin-unicorn": "^47.0.0",
 		"eslint-plugin-unused-imports": "^2.0.0",
+		"fake-indexeddb": "^5.0.2",
```
Comment (ce-bo) [inline visualization/package.json:125]: Just use v5.0.0. It will work with Node 16 https://github.com/dumbmatter/fakeIndexedDB/releases

**Comment by ce-bo** [comment]
Sorry, I oversee the information in the PR. I was blind.

File: visualization/package.json (hunk 1 above)
Comment (ce-bo) [reply visualization/package.json:125]: Sorry thats wrong. Use latest version 4.

File: visualization/app/codeCharta/ui/ribbonBar/ribbonBar.e2e.ts [hunk 2]
```diff
@@ -22,6 +22,10 @@ describe("RibbonBar", () => {
 		await goto()
 	})
 
+	afterEach(async () => {
+		await clearIndexedDB()
+	})
+
 	/*it("hovering over a folder should display the sum of metric of all children", async () => {
```
Comment (Nereboss) [inline visualization/app/codeCharta/ui/ribbonBar/ribbonBar.e2e.ts:29]: should this commented out code be removed? it has been there since 2021

File: visualization/app/codeCharta/util/indexedDB/indexedDBWriter.spec.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+			await writeCcState(defaultState)
+
+			const database = await openDB(DB_NAME, DB_VERSION)
+			const transaction = database.transaction(CCSTATE_STORE_NAME, "readonly")
+			const store = transaction.objectStore(CCSTATE_STORE_NAME)
+			const result = await store.get(CCSTATE_STATE_ID)
+
+			expect(result.state).toEqual(defaultState)
+
+			database.close()
```
Comment (Nereboss) [inline visualization/app/codeCharta/util/indexedDB/indexedDBWriter.spec.ts:30]: here the database is closed after the expect, in the tests below it is closed before the expect, both should be handled equally

**Review by Nereboss — COMMENTED** [review]
Further things found while testing:
- The camera position is currently not saved. Is this intended?
- global settings do not saved when error occurs
- turning off camera reset in global settings behaves differently that on main
- In the file/node explorer the states of the pin/sort order buttons are saved but not which folders are open/closed. Should we save this aswell?
    - Bug: The state of the pin button is saved but the menu does not stay pinned through a reload
- Before the reload save is released we should have a button to reset to the default values. Should this be done in this PR?
- Should we save all maps the user ever loaded? Does this have an impact on performance for many large maps?
- Now that the defaults for size, height and color do not appear after every reload, should we add (default) behind them?

**Comment by fritschldwg** [comment]
> Further things found while testing:
> 
> * The camera position is currently not saved. Is this intended?
→ This is a bug not introduced through this PR

> * global settings do not saved when error occurs
→ Fixed

> * turning off camera reset in global settings behaves differently that on main
→ Fixed

> * In the file/node explorer the states of the pin/sort order buttons are saved but not which folders are open/closed. Should we save this aswell?
> * Bug: The state of the pin button is saved but the menu does not stay pinned through a reload
→ Upon taking a quick look at the code, it would be a big effort to synchronize the file-explorer behaviour with the store. Imo this is not a tradeworth worth taking.

> * Before the reload save is released we should have a button to reset to the default values. Should this be done in this PR?
→ Openend a new issue regarding a reset-button

> * Should we save all maps the user ever loaded? Does this have an impact on performance for many large maps?
→ I did a couple quick performance checks in both Chrome and Firefox and the performance seems to be similar to the one before this PR

> * Now that the defaults for size, height and color do not appear after every reload, should we add (default) behind them?

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 4]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+				this.applyAppSettings(savedCcState.appSettings)
+				this.loadFileService.loadFiles(savedNameDataPairs)
+				this.store.dispatch(setFiles({ value: savedFileStates }))
+				this.applyAllSettings(savedCcState)
+				this.setRenderStateFromUrl()
+			} else {
+				this.applyAllSettings(savedCcState)
+				this.loadFileService.loadFiles(urlNameDataPairs)
+				this.setRenderStateFromUrl()
+			}
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:107]: This could be extracted into to properly named methods to make the code more readable.

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 5]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+				this.setRenderStateFromUrl()
+				return
+			}
+
+			const savedFileStates = savedCcState.files
+			const savedNameDataPairs = savedFileStates.map(fileState => getNameDataPair(fileState.file))
+			const urlNameDataPairCheckSums = urlNameDataPairs.map(urlNameDataPair => urlNameDataPair.content.fileChecksum)
+			const savedNameDataPairCheckSums = savedNameDataPairs.map(savedNameDataPair => savedNameDataPair.content.fileChecksum)
+			if (stringify(urlNameDataPairCheckSums) === stringify(savedNameDataPairCheckSums)) {
+				this.applyAppSettings(savedCcState.appSettings)
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:98]: Why are the appSettings applied before calling apllyAllSettings afterwards in line 101?

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 6]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+				this.loadFileService.loadFiles(savedNameDataPairs)
+				this.store.dispatch(setFiles({ value: savedFileStates }))
+				this.applyAllSettings(savedCcState)
+				this.setRenderStateFromUrl()
+			} else {
+				this.applyAllSettings(savedCcState)
+				this.loadFileService.loadFiles(urlNameDataPairs)
+				this.setRenderStateFromUrl()
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:106]: This can be called outside of the else

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 7]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
-			this.dialog.open(ErrorDialogComponent, {
-				data: { title, message }
-			})
+
+			const savedFileStates = savedCcState.files
+			const savedNameDataPairs = savedFileStates.map(fileState => getNameDataPair(fileState.file))
+			this.applyAppSettings(savedCcState.appSettings)
+			this.loadFileService.loadFiles(savedNameDataPairs)
+			this.store.dispatch(setFiles({ value: savedFileStates }))
+			this.applyAllSettings(savedCcState)
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:135]: More or less duplicated code (see line 98)

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 8]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+		} catch (error) {
+			await this.handleErrorLoadFilesFromIndexedDB(error as Error)
+		}
+	}
+
+	private async handleErrorLoadFilesFromIndexedDB(error: Error) {
+		if ((error as Error).message !== NO_FILES_LOADED_ERROR_MESSAGE) {
+			const title = "File(s) could not be loaded from indexeddb. Loaded sample files instead."
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:143]: Can we reframe this a little bit to express that previously loaded files and settings could not be restored?

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 9]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+			const missingFileSettings = this.applyFileSettings(savedFileSettings)
+			missingPropertiesInSavedCcState.push(...missingFileSettings)
+		}
+		if (savedDynamicSettings) {
+			const missingDynamicSettings = this.applyDynamicSettings(savedDynamicSettings)
+			missingPropertiesInSavedCcState.push(...missingDynamicSettings)
+		}
+		if (savedAppSettings) {
+			const missingAppSettings = this.applyAppSettings(savedAppSettings)
+			missingPropertiesInSavedCcState.push(...missingAppSettings)
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:165]: If something is not in the savedState, it is missing... ok. But shouldn't we simply add the missing properties to the savedState, so that it is up to date instead of showing an error? Maybe I did not understand it yet.

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 10]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+				break
+			case "isColorMetricLinkedToHeightMetric":
+				this.store.dispatch(setIsColorMetricLinkedToHeightMetricAction({ value }))
+				break
+			case "enableFloorLabels":
+				this.store.dispatch(setEnableFloorLabels({ value }))
+				break
+			default: {
+				const exhaustiveCheck: never = key
+				throw new Error(`Unhandled key: ${exhaustiveCheck}`)
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:373]: Throwing errors at this position will most likely prevent to restore old state, right?
Maybe the approach to just ignore unknown actions would be more beneficial for the users.

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 11]
```diff
@@ -25,50 +71,351 @@ export class LoadInitialFileService {
...
+			case "resetCameraIfNewFileIsLoaded":
+				this.store.dispatch(setResetCameraIfNewFileIsLoaded({ value }))
+				break
+			case "isLoadingMap":
+				this.store.dispatch(setIsLoadingMap({ value }))
+				break
+			case "isLoadingFile":
+				this.store.dispatch(setIsLoadingFile({ value }))
+				break
+			case "sortingOrderAscending":
```
[the remark this answers is not in the document]
Comment (fritschldwg) [reply visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:338]: I added a comment explaining that those properties not applied on purpose. If we were to apply the sorting-order and the pin-state we would also have to save which folders were opened / collapsed in the file-explorer. Because this would have meant a lot of effort, we decided to ignore all propeties belonging to the file-explorer.

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts (hunk 10 above)
Comment (fritschldwg) [reply visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:373]: This is actually just a ts-lint "hack" to make sure all possible values are checked in the switch statement. This will not  throw an error during runtime.

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts (hunk 9 above)
Comment (fritschldwg) [reply visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:165]: I implemented error-handling to ensure changes in the format of the cc.json or our internal settings are handled gracefully. If you've previously blacklisted items in CodeCharta and we've since updated how we store those settings, there will be a notification informing you that all settings were restored correctly, except the blacklisted items (as you would have expected).

File: visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts [hunk 12]
```diff
@@ -95,21 +95,44 @@ export class LoadInitialFileService {
...
+		}
+	}
+
+	private applySettingsFromSavedState(savedCcState: CcState, urlNameDataPairs: NameDataPair[]) {
+		this.applyAllSettings(savedCcState)
+		this.loadFileService.loadFiles(urlNameDataPairs)
+	}
+
+	private showErrorDialogForMissingProperties(missingPropertiesInSavedCcState) {
+		const title = "All properties expect the following were restored successfully."
```
Comment (ce-bo) [inline visualization/app/codeCharta/services/loadInitialFile/loadInitialFile.service.ts:131]: Is this shown to the userse? I would rephrase it a little bit, please.
I would give a understandable summary like "The previous state could not be restored after reloading the page."
Then I would give more details about it in a separate paragraph or collpased list with "see more details".

File: visualization/app/index.html [hunk 13]
```diff
@@ -5,7 +5,7 @@
 		<meta
 			charset="utf-8"
 			http-equiv="Content-Security-Policy"
-			content="default-src 'self' 'unsafe-inline' 'unsafe-eval'; 
+			content="default-src 'self' 'unsafe-inline' 'unsafe-eval' localhost:3001;
```
Comment (ce-bo) [inline visualization/app/index.html:8]: Should this be pushed?

File: visualization/app/codeCharta/ui/resetStateButton/confirmResetStateDialog/confirmResetStateDialog.component.html [hunk 14]
```diff
@@ -0,0 +1,8 @@
+<mat-toolbar class="header" xmlns="http://www.w3.org/1999/html">
+	<h2>Confirm reset map to default</h2>
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/resetStateButton/confirmResetStateDialog/confirmResetStateDialog.component.html]: Would you please add a short explanation what is going to be reset?
E.g. "Uploaded Maps and settings are reset to defaults"
